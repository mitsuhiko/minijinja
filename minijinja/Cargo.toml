[package]
name = "minijinja"
version = "2.0.3"
edition = "2021"
license = "Apache-2.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
description = "a powerful template engine for Rust with minimal dependencies"
homepage = "https://github.com/mitsuhiko/minijinja"
repository = "https://github.com/mitsuhiko/minijinja"
keywords = ["jinja", "jinja2", "templates"]
readme = "README.md"
rust-version = "1.61"
exclude = ["tests"]

[package.metadata.docs.rs]
features = ["loader", "json", "urlencode", "custom_syntax", "fuel"]
rustdoc-args = ["--cfg", "docsrs", "--html-in-header", "doc-header.html"]

[features]
default = [
    "builtins",
    "debug",
    "deserialization",
    "macros",
    "multi_template",
    "adjacent_loop_items",
    "std_collections",
]

# API features
preserve_order = ["indexmap"]
deserialization = []
debug = []
loader = ["self_cell", "memo-map"]
unicode = ["unicode-ident", "unicase"]
custom_syntax = ["dep:aho-corasick"]
std_collections = []

# Speedups
key_interning = []
speedups = ["v_htmlescape"]

# Engine Features
builtins = []
macros = []
multi_template = []
adjacent_loop_items = []
fuel = []

# Extra Filters
json = ["serde_json"]
urlencode = ["percent-encoding"]

# Internal Features that should not be used
internal_debug = []
unstable_machinery = ["internal_debug"]
unstable_machinery_serde = ["unstable_machinery", "serde/derive"]

[dependencies]
aho-corasick = { version = "1.0", default-features = false, optional = true }
serde = "1.0.130"
v_htmlescape = { version = "0.15.8", optional = true }
self_cell = { version = "1.0.4", optional = true }
serde_json = { version = "1.0.68", optional = true }
percent-encoding = { version = "2.2.0", optional = true }
indexmap = { version = "1.9.0", optional = true }
memo-map = { version = "0.3.1", optional = true }
unicode-ident = { version = "1.0.5", optional = true }
unicase = { version = "2.6.0", optional = true }
stacker = { version = "0.1.15", optional = true }

[dev-dependencies]
insta = { version = "1.38.0", features = ["glob", "serde"] }
serde = { version = "1.0.130", features = ["derive"] }
serde_json = "1.0.68"
similar-asserts = "1.4.2"
