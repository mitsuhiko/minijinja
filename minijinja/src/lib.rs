//! <div align=center>
//!   <img src="https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo.png" alt="" width=320>
//!   <p><strong>MiniJinja: a powerful template engine for Rust with minimal dependencies</strong></p>
//! </div>
//!
//! MiniJinja is a powerful but minimal dependency template engine for Rust which
//! is based on the syntax and behavior of the
//! [Jinja2](https://jinja.palletsprojects.com/) template engine for Python.  It's
//! implemented on top of [`serde`].  The goal is to be able to render a large
//! chunk of the Jinja2 template ecosystem from Rust with a minimal engine and to
//! leverage an already existing ecosystem of editor integrations.
//!
//! ```jinja
//! {% for user in users %}
//!   <li>{{ user.name }}</li>
//! {% endfor %}
//! ```
//!
//! You can play with MiniJinja online [in the browser
//! playground](https://mitsuhiko.github.io/minijinja-playground/) powered by a
//! WASM build of MiniJinja.
//!
//! # Why MiniJinja
//!
//! MiniJinja by its name wants to be a good default choice if you need a little
//! bit of templating with minimal dependencies.  It has the following goals:
//!
//! * Well documented, compact API
//! * Minimal dependencies, reasonable compile times and decent runtime performance
//! * Stay close as possible to Jinja2
//! * Support for expression evaluation
//! * Support for all `serde` compatible types
//! * Excellent test coverage
//! * Support for dynamic runtime objects with methods and dynamic attributes
//!
//! # Template Usage
//!
//! To use MiniJinja one needs to create an [`Environment`] and populate it with
//! templates.  Afterwards templates can be loaded and rendered.  To pass data
//! one can pass any serde serializable value.  The [`context!`] macro can be
//! used to quickly construct a template context:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let mut env = Environment::new();
//! env.add_template("hello", "Hello {{ name }}!").unwrap();
//! let tmpl = env.get_template("hello").unwrap();
//! println!("{}", tmpl.render(context!(name => "John")).unwrap());
//! ```
//!
//! ```plain
//! Hello John!
//! ```
//!
//! For super trivial cases where you need to render a string once, you can
//! also use the [`render!`] macro which acts a bit like a replacement
//! for the [`format!`] macro.
//!
//! # Expression Usage
//!
//! MiniJinja — like Jinja2 — allows to be used as expression language.  This can be
//! useful to express logic in configuration files or similar things.  For this
//! purpose the [`Environment::compile_expression`] method can be used.  It returns
//! an expression object that can then be evaluated, returning the result:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let env = Environment::new();
//! let expr = env.compile_expression("number < 42").unwrap();
//! let result = expr.eval(context!(number => 23)).unwrap();
//! assert_eq!(result.is_true(), true);
//! ```
//!
//! This becomes particularly powerful when [dynamic objects](crate::value::Object) are
//! exposed to templates.
//!
//! # Custom Filters
//!
//! MiniJinja lets you register functions as filter functions (see
//! [`Filter`](crate::filters::Filter)) with the engine.  These can then be
//! invoked directly from the template:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let mut env = Environment::new();
//! env.add_filter("repeat", str::repeat);
//! env.add_template("hello", "{{ 'Na '|repeat(3) }} {{ name }}!").unwrap();
//! let tmpl = env.get_template("hello").unwrap();
//! println!("{}", tmpl.render(context!(name => "Batman")).unwrap());
//! ```
//!
//! ```plain
//! Na Na Na Batman!
//! ```
//!
//! # Learn more
//!
//! - [`Environment`]: the main API entry point.  Teaches you how to configure the environment.
//! - [`Template`]: the template object API.  Shows you how templates can be rendered.
//! - [`syntax`]: provides documentation of the template engine syntax.
//! - [`filters`]: teaches you how to write custom filters and to see the list of built-in filters.
//! - [`tests`]: teaches you how to write custom test functions and to see the list of built-in tests.
//! - [`functions`]: teaches how to write custom functions and to see the list of built-in functions.
//! - For auto reloading use the [`minijinja-autoreload`](https://docs.rs/minijinja-autoreload) crate.
//!
//! Additionally there is an [list of examples](https://github.com/mitsuhiko/minijinja/tree/main/examples)
//! with many different small example programs on GitHub to explore.
//!
//! # Error Handling
//!
//! MiniJinja tries to give you good errors out of the box.  However if you use includes or
//! template inheritance your experience will improve greatly if you ensure to render chained
//! errors.  For more information see [`Error`] with an example.
//!
//! # Size and Compile Times
//!
//! MiniJinja attempts to compile fast so it can be used as a sensible template engine choice
//! when compile times matter.  Because of this it's internally modular so unnecessary bits and
//! pieces can be removed.  The default set of features tries to strike a balance but in
//! situations where only a subset is needed (eg: `build.rs`) all default features can be
//! be disabled.
//!
//! # Optional Features
//!
//! MiniJinja comes with a lot of optional features, some of which are turned on by
//! default.  If you plan on using MiniJinja in a library, please consider turning
//! off all default features and to opt-in explicitly into the ones you actually
//! need.
//!
//! <details><summary><strong style="cursor: pointer">Configurable Features</strong></summary>
//!
//! To cut down on size of the engine some default functionality can be removed:
//!
//! - **Engine Features:**
//!
//!   - `builtins`: if this feature is removed the default filters, tests and
//!     functions are not implemented.
//!   - `macros`: when removed the `{% macro %}` tag is not included.
//!   - `multi_template`: when removed the templates related to imports and extends
//!     are removed (`{% from %}`, `{% import %}`, `{% include %}`, and `{% extends %}`
//!     as well as `{% block %}`).
//!   - `adjacent_loop_items`: when removed the `previtem` and `nextitem` attributes of
//!     the `loop` object no longer exist.  Removing this feature can provide faster
//!     template execution when a lot of loops are involved.
//!   - `unicode`: when added unicode identifiers are supported.  Without this features
//!     only ASCII identifiers can be used for variable names and attributes.
//!
//! - **Rust Functionality:**
//!
//!   - `debug`: if this feature is removed some debug functionality of the engine is
//!     removed as well.  This mainly affects the quality of error reporting.
//!   - `deserialization`: when removed this disables deserialization support for
//!     the [`Value`](crate::value::Value) type.
//!
//! There are some additional features that can be enabled:
//!
//! - `fuel`: enables the `fuel` feature which makes the engine track fuel consumption which
//!   can be used to better protect against expensive templates.
//! - `source`: enables the `Source` type which helps with dynamic loading of templates.
//! - `speedups`: enables all speedups, in particular it turns on the `v_htmlescape` dependency
//!   for faster HTML escapling.
//! - `json`: When enabled the `tojson` filter is added as builtin filter as well as
//!   the ability to auto escape via `AutoEscape::Json`.
//! - `urlencode`: When enabled the `urlencode` filter is added as builtin filter.
//! - `preserve_order`: When enable the internal value implementation uses an indexmap
//!   which preserves the original order of maps and structs.
//! - `key_interning`: if this feature is enabled the automatic string interning in
//!   the value type is enabled.  This feature used to be turned on by default but
//!   has negative performance effects in newer versions of MiniJinja since a lot of
//!   the previous uses of key interning are no longer needed.  Enabling it however
//!   cuts down on memory usage slightly in certain scenarios by interning all string
//!   keys used in dynamic map values.
//!
//! </details>
#![allow(clippy::cognitive_complexity)]
#![allow(clippy::get_first)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![deny(missing_docs)]
#![doc(html_logo_url = "https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo-square.png")]

#[macro_use]
mod macros;

mod compiler;
mod defaults;
mod environment;
mod error;
mod expression;
mod key;
mod output;
mod template;
mod utils;
mod vm;

pub mod filters;
pub mod functions;
pub mod syntax;
pub mod tests;
pub mod value;

#[cfg(any(test, feature = "testutils"))]
#[cfg_attr(docsrs, doc(cfg(feature = "testutils")))]
pub mod testutils;

#[cfg(feature = "source")]
mod source;

#[cfg(feature = "debug")]
mod debug;

pub use self::defaults::{default_auto_escape_callback, escape_formatter};
pub use self::environment::Environment;
pub use self::error::{Error, ErrorKind};
pub use self::expression::Expression;
pub use self::output::Output;
pub use self::template::Template;
pub use self::utils::{AutoEscape, HtmlEscape, UndefinedBehavior};

#[cfg(feature = "source")]
pub use self::source::Source;

pub use self::macros::__context;
pub use self::vm::State;

/// This module gives access to the low level machinery.
///
/// This module is only provided by the `unstable_machinery` feature and does not
/// have a stable interface.  It mostly exists for internal testing purposes and
/// for debugging.
#[cfg(feature = "unstable_machinery")]
#[cfg_attr(docsrs, doc(cfg(feature = "unstable_machinery")))]
pub mod machinery {
    #![allow(missing_docs)]
    pub use crate::compiler::ast;
    pub use crate::compiler::codegen::CodeGenerator;
    pub use crate::compiler::instructions::{Instruction, Instructions};
    pub use crate::compiler::lexer::tokenize;
    pub use crate::compiler::parser::parse;
    pub use crate::compiler::tokens::{Span, Token};
    pub use crate::template::CompiledTemplate;
    pub use crate::vm::Vm;

    use crate::Output;

    /// Creates an [`Output`] that writes into a string.
    pub fn make_string_output(s: &mut String) -> Output<'_> {
        Output::with_string(s)
    }
}
