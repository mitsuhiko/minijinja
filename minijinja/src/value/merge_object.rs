use std::collections::BTreeSet;
use std::sync::Arc;

use crate::value::object::StructObject;
use crate::value::Value;

/// Utility struct used by [`context!`](crate::context) to merge
/// multiple values.
pub struct MergeObject(pub Vec<Value>);

impl StructObject for MergeObject {
    fn get_field(&self, field: &str) -> Option<Value> {
        for val in &self.0 {
            match val.get_attr(field) {
                Ok(val) if !val.is_undefined() => return Some(val),
                _ => {}
            }
        }
        None
    }

    fn fields(&self) -> Vec<Arc<str>> {
        let mut seen = BTreeSet::new();
        let mut rv = Vec::new();
        for val in &self.0 {
            if let Ok(iter) = val.try_iter() {
                for item in iter {
                    let s: Result<Arc<str>, _> = item.try_into();
                    if let Ok(s) = s {
                        if !seen.contains(&s) {
                            seen.insert(s.clone());
                            rv.push(s);
                        }
                    }
                }
            }
        }
        rv
    }
}
