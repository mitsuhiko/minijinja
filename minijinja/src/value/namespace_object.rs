use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};

use crate::value::{StructObject, Value};

/// This object exists for the `namespace` function.
///
/// It's special in that it behaves like a dictionary in many ways but it's the only
/// object that can be used with `{% set %}` assignments.  This is used internally
/// in the vm via downcasting.
#[derive(Default)]
pub(crate) struct Namespace {
    data: Mutex<BTreeMap<Arc<str>, Value>>,
}

impl StructObject for Namespace {
    fn get_field(&self, name: &str) -> Option<Value> {
        self.data.lock().unwrap().get(name).cloned()
    }

    fn fields(&self) -> Vec<Arc<str>> {
        self.data.lock().unwrap().keys().cloned().collect()
    }

    fn field_count(&self) -> usize {
        self.data.lock().unwrap().len()
    }
}

impl Namespace {
    pub(crate) fn set_field(&self, key: &str, value: Value) {
        self.data.lock().unwrap().insert(key.into(), value);
    }
}
