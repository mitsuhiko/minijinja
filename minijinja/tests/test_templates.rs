#![cfg(all(
    feature = "multi_template",
    feature = "macros",
    feature = "builtins",
    feature = "adjacent_loop_items"
))]
use std::collections::BTreeMap;
use std::fmt::Write;
use std::sync::Arc;
use std::{env, fs};

use insta::assert_snapshot;
use minijinja::value::{Enumerator, Object, ObjectRepr, Value};
use minijinja::{context, render, Environment, Error, ErrorKind, State};

use similar_asserts::assert_eq;

#[test]
fn test_vm() {
    let mut refs = Vec::new();
    insta::glob!("inputs/refs/*", |entry| {
        let filename = entry.file_name().unwrap();
        let filename = filename.to_str().unwrap();
        if filename.ends_with(".txt") || filename.ends_with(".html") {
            let source = fs::read_to_string(entry).unwrap();
            refs.push((entry.to_path_buf(), source));
        }
    });

    insta::glob!("inputs/*", |path| {
        if !path.metadata().unwrap().is_file() {
            return;
        }
        let filename = path.file_name().unwrap().to_str().unwrap();
        let contents = std::fs::read_to_string(path).unwrap();
        let mut iter = contents.splitn(2, "\n---\n");
        let mut env = Environment::new();
        let ctx: serde_json::Value = serde_json::from_str(iter.next().unwrap()).unwrap();

        for (path, source) in &refs {
            let ref_filename = path.file_name().unwrap().to_str().unwrap();
            env.add_template(ref_filename, source).unwrap();
        }

        let content = iter.next().unwrap();
        let rendered = if let Err(err) = env.add_template(filename, content) {
            let mut rendered = format!("!!!SYNTAX ERROR!!!\n\n{err:#?}\n\n");
            writeln!(rendered, "{err:#}").unwrap();
            rendered
        } else {
            let template = env.get_template(filename).unwrap();

            match template.render(&ctx) {
                Ok(mut rendered) => {
                    rendered.push('\n');
                    rendered
                }
                Err(err) => {
                    let mut rendered = format!("!!!ERROR!!!\n\n{err:#?}\n\n");

                    writeln!(rendered, "{err:#}").unwrap();
                    let mut err = &err as &dyn std::error::Error;
                    while let Some(next_err) = err.source() {
                        writeln!(rendered).unwrap();
                        writeln!(rendered, "caused by: {next_err:#}").unwrap();
                        err = next_err;
                    }

                    rendered
                }
            }
        };

        insta::with_settings!({
            info => &ctx,
            description => content.trim_end(),
            omit_expression => true
        }, {
            insta::assert_snapshot!(&rendered);
        });
    });
}

#[test]
fn test_vm_block_fragments() {
    let mut refs = Vec::new();
    insta::glob!("fragment-inputs/refs/*", |entry| {
        let filename = entry.file_name().unwrap();
        let filename = filename.to_str().unwrap();
        if filename.ends_with(".txt") || filename.ends_with(".html") {
            let source = fs::read_to_string(entry).unwrap();
            refs.push((entry.to_path_buf(), source));
        }
    });

    insta::glob!("fragment-inputs/*", |path| {
        if !path.metadata().unwrap().is_file() {
            return;
        }
        let filename = path.file_name().unwrap().to_str().unwrap();
        let contents = std::fs::read_to_string(path).unwrap();
        let mut iter = contents.splitn(2, "\n---\n");
        let mut env = Environment::new();
        let ctx: serde_json::Value = serde_json::from_str(iter.next().unwrap()).unwrap();

        for (path, source) in &refs {
            let ref_filename = path.file_name().unwrap().to_str().unwrap();
            env.add_template(ref_filename, source).unwrap();
        }

        let content = iter.next().unwrap();
        let rendered = if let Err(err) = env.add_template(filename, content) {
            let mut rendered = format!("!!!SYNTAX ERROR!!!\n\n{err:#?}\n\n");
            writeln!(rendered, "{err:#}").unwrap();
            rendered
        } else {
            let template = env.get_template(filename).unwrap();

            match template
                .eval_to_state(&ctx)
                .and_then(|mut x| x.render_block("fragment"))
            {
                Ok(mut rendered) => {
                    rendered.push('\n');
                    rendered
                }
                Err(err) => {
                    let mut rendered = format!("!!!ERROR!!!\n\n{err:#?}\n\n");

                    writeln!(rendered, "{err:#}").unwrap();
                    let mut err = &err as &dyn std::error::Error;
                    while let Some(next_err) = err.source() {
                        writeln!(rendered).unwrap();
                        writeln!(rendered, "caused by: {next_err:#}").unwrap();
                        err = next_err;
                    }

                    rendered
                }
            }
        };

        insta::with_settings!({
            info => &ctx,
            description => content.trim_end(),
            omit_expression => true
        }, {
            insta::assert_snapshot!(&rendered);
        });
    });
}

#[test]
fn test_custom_filter() {
    fn test_filter(_: &State, value: String) -> Result<String, Error> {
        Ok(format!("[{value}]"))
    }

    let mut ctx = BTreeMap::new();
    ctx.insert("var", 42);

    let mut env = Environment::new();
    env.add_filter("test", test_filter);
    env.add_template("test", "{{ var|test }}").unwrap();
    let tmpl = env.get_template("test").unwrap();
    let rv = tmpl.render(&ctx).unwrap();
    assert_eq!(rv, "[42]");
}

#[test]
fn test_items_and_dictsort_with_structs() {
    #[derive(Debug, Clone)]
    struct MyStruct;

    impl Object for MyStruct {
        fn get_value(self: &Arc<Self>, name: &Value) -> Option<Value> {
            match name.as_str()? {
                "a" => Some(Value::from("A")),
                "b" => Some(Value::from("B")),
                _ => None,
            }
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Str(&["b", "a"])
        }
    }

    insta::assert_snapshot!(
        minijinja::render!("{{ x|items }}", x => Value::from_object(MyStruct)),
        @r###"[["b", "B"], ["a", "A"]]"###
    );
    insta::assert_snapshot!(
        minijinja::render!("{{ x|dictsort }}", x => Value::from_object(MyStruct)),
        @r###"[["a", "A"], ["b", "B"]]"###
    );
}

#[test]
fn test_urlencode_with_struct() {
    #[derive(Debug, Clone)]
    struct MyStruct;

    impl Object for MyStruct {
        fn get_value(self: &Arc<Self>, name: &Value) -> Option<Value> {
            match name.as_str() {
                Some("a") => Some(Value::from("a 1")),
                Some("b") => Some(Value::from("b 2")),
                _ => None,
            }
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Str(&["a", "b"])
        }
    }

    insta::assert_snapshot!(
        minijinja::render!("{{ x|urlencode }}", x => Value::from_object(MyStruct)),
        @"a=a%201&b=b%202"
    );
}

#[test]
fn test_single() {
    let mut env = Environment::new();
    env.add_template("simple", "Hello {{ name }}!").unwrap();
    let tmpl = env.get_template("simple").unwrap();
    let rv = tmpl.render(context!(name => "Peter")).unwrap();
    assert_eq!(rv, "Hello Peter!");
}

#[test]
fn test_values_scientific_notation() {
    let mut env = Environment::new();
    env.add_template("sci1", "VALUE = {{ value or -12.4E-4 }}")
        .unwrap();
    let tmpl = env.get_template("sci1").unwrap();
    let rv = tmpl.render(context!(value => -12.4E-3)).unwrap();
    assert_eq!(rv, "VALUE = -0.0124");
    let rv = tmpl.render(context!());
    // assert_eq!(rv, "VALUE = -0.00124");
    assert!(rv.is_ok());

    env.add_template("sci2", "VALUE = {{ value or 1.4E4 }}")
        .unwrap();
    let tmpl = env.get_template("sci2").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci3", "VALUE = {{ value or 1.4e+4}}")
        .unwrap();
    let tmpl = env.get_template("sci3").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci4", "VALUE = {{ 1.4+4}}").unwrap();
    let tmpl = env.get_template("sci4").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci5", "VALUE = {{ 1.4+1E-1}}").unwrap();
    let tmpl = env.get_template("sci5").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci6", "VALUE = {{ 1.0E0+1.0}}").unwrap();
    let tmpl = env.get_template("sci6").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());
}

#[test]
fn test_auto_escaping() {
    let mut env = Environment::new();
    env.add_template("index.html", "{{ var }}").unwrap();
    #[cfg(feature = "json")]
    {
        env.add_template("index.js", "{{ var }}").unwrap();
    }
    env.add_template("index.txt", "{{ var }}").unwrap();

    // html
    let tmpl = env.get_template("index.html").unwrap();
    let rv = tmpl.render(context!(var => "<script>")).unwrap();
    insta::assert_snapshot!(rv, @"&lt;script&gt;");

    // JSON
    #[cfg(feature = "json")]
    {
        use minijinja::value::Value;
        let tmpl = env.get_template("index.js").unwrap();
        let rv = tmpl.render(context!(var => "foo\"bar'baz")).unwrap();
        insta::assert_snapshot!(rv, @r###""foo\"bar'baz""###);
        let rv = tmpl
            .render(context!(var => [Value::from(true), Value::from("<foo>"), Value::from(())]))
            .unwrap();
        insta::assert_snapshot!(rv, @r###"[true,"<foo>",null]"###);
    }

    // Text
    let tmpl = env.get_template("index.txt").unwrap();
    let rv = tmpl.render(context!(var => "foo\"bar'baz")).unwrap();
    insta::assert_snapshot!(rv, @r###"foo"bar'baz"###);
}

#[test]
fn test_loop_changed() {
    let rv = minijinja::render!(
        r#"
        {%- for i in items -%}
          {% if loop.changed(i) %}{{ i }}{% endif %}
        {%- endfor -%}
        "#,
        items => vec![1, 1, 1, 2, 3, 4, 4, 5],
    );
    assert_eq!(rv, "12345");
}

// ideally this would work, but unfortunately the way serde flatten works makes it
// impossible for us to support with the internal optimizations in the value model.
// see https://github.com/mitsuhiko/minijinja/issues/222
#[derive(Debug, serde::Serialize)]
struct Bad {
    a: i32,
    #[serde(flatten)]
    more: Value,
}

#[test]
#[should_panic = "can only flatten structs and maps"]
fn test_flattening() {
    let ctx = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let env = Environment::new();
    env.render_str("{{ debug() }}", ctx).unwrap();
}

#[test]
fn test_flattening_sub_item_good() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(bad, good => "good");
    let env = Environment::new();

    // we are not touching a bad value, so we are good
    let rv = env.render_str("{{ good }}", ctx).unwrap();
    assert_eq!(rv, "good");
}

#[test]
#[should_panic = "can only flatten structs and maps"]
fn test_flattening_sub_item_bad_lookup() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(bad, good => "good");
    let env = Environment::new();

    // resolving an invalid value will fail
    env.render_str("{{ bad }}", ctx).unwrap();
}

#[test]
fn test_flattening_sub_item_bad_attr() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(good => context!(bad));
    let env = Environment::new();

    // resolving an invalid value will fail, even in an attribute lookup
    let err = env
        .render_str("{% if good.bad %}...{% endif %}", ctx)
        .unwrap_err();
    assert_eq!(err.kind(), ErrorKind::BadSerialization);
    assert_eq!(
        err.detail(),
        Some("can only flatten structs and maps (got an enum)")
    );
}

#[test]
fn test_flattening_sub_item_shielded_print() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(good => context!(bad));
    let env = Environment::new();

    // this on the other hand is okay
    let value = env.render_str("{{ good }}", ctx).unwrap();
    assert_eq!(
        value,
        r#"{"bad": <invalid value: could not serialize to value: can only flatten structs and maps (got an enum)>}"#
    );
}

#[test]
#[cfg(feature = "custom_syntax")]
fn test_custom_syntax() {
    let mut env = Environment::new();
    env.set_syntax(
        minijinja::syntax::SyntaxConfig::builder()
            .block_delimiters("{", "}")
            .variable_delimiters("${", "}")
            .comment_delimiters("{*", "*}")
            .build()
            .unwrap(),
    );

    // this on the other hand is okay
    let value = env
        .render_str("{for x in range(3)}${x}{endfor}{* nothing *}", ())
        .unwrap();
    assert_eq!(value, r"012");
}

#[test]
fn test_undeclared_variables() {
    let mut env = Environment::new();
    env.add_template(
        "demo",
        "{% set x = foo %}{{ x }}{{ bar.baz }}{{ bar.blub }}",
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();
    let undeclared = tmpl.undeclared_variables(false);
    assert_eq!(
        undeclared,
        ["foo", "bar"].into_iter().map(|x| x.to_string()).collect()
    );
    let undeclared = tmpl.undeclared_variables(true);
    assert_eq!(
        undeclared,
        ["foo", "bar.baz", "bar.blub"]
            .into_iter()
            .map(|x| x.to_string())
            .collect()
    );
}

#[test]
fn test_undeclared_variables_bug() {
    // see https://github.com/mitsuhiko/minijinja/issues/484
    let mut env = Environment::new();
    env.add_template(
        "demo",
        r#"
        {% for item in content.products.items() %}
          {{ foo.bar }}
          {{ item.name }}  // this used to cause a loop
        {% endfor %}
    "#,
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();
    let undeclared = tmpl.undeclared_variables(true);
    assert_eq!(
        undeclared,
        ["foo.bar", "content.products.items"]
            .into_iter()
            .map(|x| x.to_string())
            .collect()
    );
}

#[test]
fn test_block_fragments() {
    let mut env = Environment::new();
    env.add_template(
        "demo",
        "I am outside the fragment{% block foo %}foo{% endblock %}So am I!",
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();

    let rv_a = tmpl.render(()).unwrap();
    let rv_b = tmpl.eval_to_state(()).unwrap().render_block("foo").unwrap();

    assert_eq!(rv_a, "I am outside the fragmentfooSo am I!");
    assert_eq!(rv_b, "foo");
}

#[test]
fn test_state() {
    let mut env = Environment::new();
    env.add_template(
        "foo.html",
        r#"
        {% set global = variable * 2 %}
        {% macro something() %}{{ global }}{% endmacro %}
        {% block baz %}[{{ global }}]{% endblock %}
    "#,
    )
    .unwrap();
    let template = env.get_template("foo.html").unwrap();
    let mut state = template
        .eval_to_state(context! {
            variable => 23
        })
        .unwrap();
    assert!(state.lookup("range").is_some());
    assert!(!state.exports().contains(&"range"));
    assert_eq!(state.lookup("global"), Some(Value::from(23 * 2)));
    assert_eq!(state.call_macro("something", &[]).unwrap(), "46");
    assert_eq!(state.render_block("baz").unwrap(), "[46]");
}

#[test]
#[allow(unused_mut)]
fn test_render_and_return_state() {
    let mut env = Environment::new();
    #[cfg(feature = "fuel")]
    {
        env.set_fuel(Some(100));
    }
    let tmpl = env
        .template_from_str("{% for x in range(3) %}Hello {{ name }}!\n{% endfor %}{% set x = 1 %}")
        .unwrap();
    let (rv, state) = tmpl
        .render_and_return_state(context! { name => "Foo" })
        .unwrap();
    assert_eq!(rv, "Hello Foo!\nHello Foo!\nHello Foo!\n");
    assert_eq!(state.lookup("x"), Some(Value::from(1)));

    #[cfg(feature = "fuel")]
    {
        assert_eq!(state.fuel_levels(), Some((26, 74)));
    }
}

#[test]
fn test_render_to_write_state() {
    let env = Environment::new();
    let tmpl = env
        .template_from_str("{% set foo = 42 %}{% macro bar() %}x{% endmacro %}root")
        .unwrap();
    let mut out = Vec::<u8>::new();
    let state = tmpl.render_to_write((), &mut out).unwrap();
    assert_eq!(String::from_utf8_lossy(&out), "root");
    assert_eq!(state.lookup("foo"), Some(Value::from(42)));
    assert_eq!(state.call_macro("bar", &[]).ok().as_deref(), Some("x"));
}

#[test]
fn test_functions() {
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> i32 { 42 })),
        @"42"
    );
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> Option<i32> { None })),
        @"none"
    );
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> Result<i32, Error> { Ok(23) })),
        @"23"
    );
}

#[test]
fn test_invalid_value_iteration() {
    let env = Environment::new();
    let mut out = Vec::<u8>::new();

    #[derive(Debug)]
    struct FailingIteration;

    impl Object for FailingIteration {
        fn repr(self: &Arc<Self>) -> ObjectRepr {
            ObjectRepr::Iterable
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Iter(Box::new((0..4).map(Value::from).chain(Some(Value::from(
                Error::new(ErrorKind::InvalidOperation, "failed during iteration"),
            )))))
        }
    }

    let t = env
        .template_from_str("{% for item in iter %}[{{ item }}]{% endfor %}")
        .unwrap();
    let err = t
        .render_to_write(
            context! { iter => Value::from_object(FailingIteration) },
            &mut out,
        )
        .unwrap_err();
    let out = String::from_utf8_lossy(&out);
    assert_eq!(out, "[0][1][2][3]");
    assert_eq!(err.kind(), ErrorKind::InvalidOperation);
    assert_eq!(err.detail(), Some("failed during iteration"));
}
