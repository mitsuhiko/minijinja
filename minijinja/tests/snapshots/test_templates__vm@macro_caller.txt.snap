---
source: minijinja/tests/test_templates.rs
description: "{% macro render_dialog(title, class='dialog') -%}\n    <div class=\"{{ class }}\">\n        <h2>{{ title }}</h2>\n        <div class=\"contents\">\n            {{ caller() }}\n        </div>\n    </div>\n{%- endmacro %}\n\n{% call render_dialog('Hello World') %}\n    This is a simple dialog rendered by using a macro and\n    a call block.\n{% endcall %}\n\ncaller: {{ render_dialog.caller }}"
info: {}
input_file: minijinja/tests/inputs/macro_caller.txt
---


<div class="dialog">
        <h2>Hello World</h2>
        <div class="contents">
            
    This is a simple dialog rendered by using a macro and
    a call block.

        </div>
    </div>

caller: true

